### PR TITLE
Quic compliance for connection termination

### PR DESCRIPTION
**Pull request by paulidale**
- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/quic/quic_wire.c [hunk 1]
```diff
@@ -366,18 +367,45 @@ int ossl_quic_wire_encode_frame_path_response(WPACKET *pkt,
...
+                                     : OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT))
         return 0;
 
-    if (!f->is_app && !WPACKET_quic_write_vlint(pkt, f->frame_type))
+    /*
+     * RFC 9000 10.2.3: Endpoints ... SHOULD use the APPLICATION_ERROR
+     * code when converting to a CONNECTION_CLOSE of type 0x1c.
+     */
+     if (!WPACKET_quic_write_vlint(pkt, is_1d ? f->error_code
+                                              : QUIC_ERR_APPLICATION_ERROR))
```
Comment (paulidale) [inline ssl/quic/quic_wire.c:386]: I'm a little concerned about this change.

For a 1c close, the codes come from the space defined in [20.1](https://datatracker.ietf.org/doc/html/rfc9000#transport-error-codes) which includes the application error code.  That section says they are applicable to 1c closes, yet 1c closes strongly suggest only sending application errors.

A better approach might be converting all non-20.1 code to application error and leaving the rest intact.  I.e. leave codes 00 - 3f and 100 -1ff unchanged.

Comment (t8m) [reply ssl/quic/quic_wire.c:386]: We're not sending arbitrary error codes with 1c close. We're sending only these from 20.1.

This change does not make sense. Also please see https://github.com/openssl/openssl/pull/21135/commits/741a3abd92e120bb57f2f3bbc89bc395c5149829

File: ssl/quic/quic_wire.c [hunk 2]
```diff
@@ -366,18 +367,45 @@ int ossl_quic_wire_encode_frame_path_response(WPACKET *pkt,
...
+    if (!is_1d && !WPACKET_quic_write_vlint(pkt, f->frame_type))
         return 0;
 
+    /*
+     * RFC 9000 10.2.3: Endpoints MUST clear the value of the Reason
+     * Phrase field ... when converting to a CONNECTION_CLOSE of type 0x1c.
+     */
+    if (!is_1d) {
+        if (!WPACKET_quic_write_vlint(pkt, 0))
+            return 0;
```
Comment (t8m) [inline ssl/quic/quic_wire.c:402]: This is wrong and should not be done on this level. See https://github.com/openssl/openssl/pull/21135/commits/741a3abd92e120bb57f2f3bbc89bc395c5149829

File: ssl/quic/quic_wire.c (hunk 1 above)
Comment (paulidale) [reply ssl/quic/quic_wire.c:386]: Actually we seem to be doing exactly this:

https://github.com/openssl/openssl/blob/c3a5ee3583efe5791dd4eb2827887264479a3ba2/ssl/quic/quic_wire.c#L370-L373

We're changing the packet type without guaranteeing that the error code etc are correct, although some things are taken care of.  I don't see how or why this piece of code is _safe_.

File: ssl/quic/quic_channel.c [hunk 3]
```diff
@@ -2798,9 +2854,20 @@ static void ch_update_idle(QUIC_CHANNEL *ch)
...
-            ossl_ms2time(ch->max_idle_timeout));
+    else {
+        /* RFC 9000 s. 10.1: Idle Timeout
+         *  To avoid excessively small idle timeout periods, endpoints
+         *  MUST increase the idle timeout period to be at least three
+         *  times the current Probe Timeout (PTO). This allows for
+         *  multiple PTOs to expire, and therefore multiple probes to
+         *  be sent and lost, prior to idle timeout.
+         */
+        OSSL_TIME pto = ossl_ackm_get_pto_duration(ch->ackm);
```
Comment (hlandau) [inline ssl/quic/quic_channel.c:2865]: :+1:

**Review by hlandau — COMMENTED** [review]
Looks good so far

File: ssl/quic/quic_channel.c [hunk 4]
```diff
@@ -480,6 +484,37 @@ ossl_quic_channel_get_terminate_cause(const QUIC_CHANNEL *ch)
     return ossl_quic_channel_is_term_any(ch) ? &ch->terminate_cause : NULL;
 }
 
+int ossl_quic_channel_can_send_closing_bytes(QUIC_CHANNEL *ch, size_t n)
```
Comment (hlandau) [inline ssl/quic/quic_channel.c:487]: This does not seem to be called from anywhere.

File: ssl/quic/quic_demux.c [hunk 5]
```diff
@@ -460,20 +460,22 @@ static int demux_recv(QUIC_DEMUX *demux)
...
 
 /* Identify the connection structure corresponding to a given URXE. */
-static QUIC_DEMUX_CONN *demux_identify_conn(QUIC_DEMUX *demux, QUIC_URXE *e)
+static QUIC_DEMUX_CONN *demux_identify_conn(QUIC_DEMUX *demux, QUIC_URXE *e,
+                                            int *pkt_too_short)
 {
     QUIC_CONN_ID dst_conn_id;
 
-    if (!demux_identify_conn_id(demux, e, &dst_conn_id))
+    if (!demux_identify_conn_id(demux, e, &dst_conn_id, pkt_too_short))
```
Comment (hlandau) [inline ssl/quic/quic_demux.c:478]: If a packet is sufficiently malformed that `ossl_quic_wire_get_pkt_hdr_dst_conn_id` fails, we may as well reject processing of it in any circumstance. So I would suggest getting rid of the changes in `quic_wire_pkt.c` and just modifying this function in DEMUX to indicate whether it's a failure of `demux_identify_conn_id` or a lookup failure. Flag can be a more general `pkt_invalid` or similar.

File: ssl/quic/quic_channel.c [hunk 6]
```diff
@@ -480,6 +484,37 @@ ossl_quic_channel_get_terminate_cause(const QUIC_CHANNEL *ch)
...
+     *      times the size of packets it receives from that address.
+     */
+    res = ch->closing_bytes_xmit + n <= ch->closing_bytes_recv * 3;
+
+    /*
+     * Attribute the bytes to the connection, if we are allowed to send them
+     * and this isn't the first closing frame.
+     */
+    if (res && ch->closing_bytes_recv != 0)
+        ch->closing_bytes_xmit += n;
```
Comment (hlandau) [inline ssl/quic/quic_channel.c:514]: It is a bit confusing to have a function whose name implies it's a non-mutating predicate but which does a mutation like this. Maybe `ossl_quic_channel_try_commit_conn_close` or something better.

There's a question as to whether this logic should live in the TXP but I can live with it being like this for now (potential issues with coupling the TXP to the RX side also).

**Review by hlandau — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_demux.c (hunk 5 above)
Comment (paulidale) [reply ssl/quic/quic_demux.c:478]: I think this is required.   As per RFC 9000 s. 10.3
> Endpoints MUST discard packets that are too small to be valid QUIC packets. To give an example, with the set of AEAD functions defined in [[QUIC-TLS](https://datatracker.ietf.org/doc/html/rfc9001)], short header packets that are smaller than 21 bytes are never valid.

Without knowing that the packet is undersized, the code in demux_process_pending_urxe() will pass it to the default packet handler for processing.  This must not occur and the packet must be forcibly dropped.

https://github.com/openssl/openssl/blob/06a0d40322e96dbba816b35f82226871f635ec5a/ssl/quic/quic_demux.c#L499-L514

I do like the idea of making this a flags argument instead.

File: ssl/quic/quic_channel.c (hunk 6 above)
Comment (paulidale) [reply ssl/quic/quic_channel.c:514]: I've moved the logic to the TXP.  It probably sits more nicely there, although it's not quite so clean IMO.

File: ssl/quic/quic_channel.c (hunk 4 above)
Comment (paulidale) [reply ssl/quic/quic_channel.c:487]: Lost somewhere. It was meant to be called from the TXP.
Anyway, refactored now so it lives in the TXP.

File: ssl/quic/quic_demux.c (hunk 5 above)
Comment (hlandau) [reply ssl/quic/quic_demux.c]: The QRX rejects any packet which is too small to be a valid encrypted packet.

Any handling in the default packet handler will begin with some specific unmarshalling which will enforce any necessary packet geometry as an implication of its operation. So we do not need to do anything in the DEMUX to meet this requirement.

File: ssl/quic/quic_channel.c [hunk 7]
```diff
@@ -2564,6 +2635,12 @@ static void ch_start_terminating(QUIC_CHANNEL *ch,
         if (force_immediate)
             ch_on_terminating_timeout(ch);
         else if (tcause->remote)
+            /*
+             * RFC 9000 s. 10.2.2 Drainging Connection State:
```
Comment (hlandau) [inline ssl/quic/quic_channel.c:2639]: nit: Draining

File: ssl/quic/quic_txp.c [hunk 8]
```diff
@@ -1640,6 +1647,46 @@ static void on_sstream_updated(uint64_t stream_id, void *arg)
...
+    /*
+     * Attribute the bytes to the connection, if we are allowed to send them
+     * and this isn't the first closing frame.
+     */
+    if (res && txp->closing_bytes_recv != 0)
+        txp->closing_bytes_xmit += n;
+    return res;
+}
+
+void txp_record_received_closing_bytes(OSSL_QUIC_TX_PACKETISER *txp, size_t n)
```
Comment (hlandau) [inline ssl/quic/quic_txp.c:1685]: Name needs qualifying (`ossl_quic_tx_packetiser`)

**Review by hlandau — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_txp.c (hunk 8 above)
Comment (paulidale) [reply ssl/quic/quic_txp.c]: Done.

**Comment by paulidale** [comment]
Remembered to push my changes this time around.

File: include/internal/quic_txp.h [hunk 9]
```diff
@@ -56,7 +56,15 @@ typedef struct ossl_quic_tx_packetiser_args_st {
...
      */
     QUIC_SSTREAM    *crypto[QUIC_PN_SPACE_NUM];
-} OSSL_QUIC_TX_PACKETISER_ARGS;
+
+    /*
+     * Counts of the number of bytes received and sent while in the closing
+     * state.
+     */
+    uint64_t                        closing_bytes_recv;
+    uint64_t                        closing_bytes_xmit;
```
Comment (hlandau) [inline include/internal/quic_txp.h:65]: These fields seem to be duplicated in the args struct unnecessarily and are never touched by the TXP. They can be removed.

**Review by hlandau — CHANGES_REQUESTED** [review]
One last change and LGTM

File: include/internal/quic_txp.h (hunk 9 above)
Comment (paulidale) [reply include/internal/quic_txp.h]: Nice catch.  Removed.

**Review by hlandau — APPROVED** [review]

**Comment by paulidale** [comment]
Ping @openssl/committers for second review please.

**Review by t-j-h — APPROVED** [review]
That took longer than expected - thanks for the RFC references which led to me re-reading some stuff ...

**Comment by paulidale** [comment]
@t-j-h reviews are taking forever currently which is an impediment.
The intention here was mostly to add comments pointing to the RFCs and confirming that what was required was done, the amount of code required was fairly small.

I split out the stateless reset effort because there is more there to be done and it involves a quite a bit more than referencing the RFCs.  It's a work in progress currently.

**Comment by paulidale** [comment]
Merged, thanks for the reviews and comments.  Well worthwhile.

**Pull request closed without merging**